RESTEasy Reactive doesn't provide converters for types like `Optional<LocalDate>`

### Describe the bug

Not sure if this is expected to work or not, but while RestEasy Reactive has support for Java Time types (like `LocalDate`) out-of-the-box, this doesn't work when wrapped in an optional (e.g. `Optional<LocalDate>`).

```
2022-09-20 11:33:03,799 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.resteasy.reactive.server.deployment.ResteasyReactiveProcessor#setupEndpoints threw an exception: java.lang.RuntimeException: java.lang.RuntimeException: Failed to process method 'org.acme.GreetingResource#hello'
        at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:313)
        at io.quarkus.resteasy.reactive.server.deployment.ResteasyReactiveProcessor.setupEndpoints(ResteasyReactiveProcessor.java:584)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:281)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at java.base/java.lang.Thread.run(Thread.java:833)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: java.lang.RuntimeException: Failed to process method 'org.acme.GreetingResource#hello'
        at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createResourceMethod(EndpointIndexer.java:693)
        at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:400)
        at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:284)
        ... 12 more
Caused by: java.lang.RuntimeException: Failed to find converter for java.time.LocalDate
        at org.jboss.resteasy.reactive.server.processor.generation.converters.GeneratedConverterIndexerExtension.extractConverterImpl(GeneratedConverterIndexerExtension.java:104)
        at org.jboss.resteasy.reactive.server.processor.ServerEndpointIndexer.extractConverter(ServerEndpointIndexer.java:507)
        at org.jboss.resteasy.reactive.server.processor.ServerEndpointIndexer.handleOptionalParam(ServerEndpointIndexer.java:357)
        at org.jboss.resteasy.reactive.server.processor.ServerEndpointIndexer.handleOptionalParam(ServerEndpointIndexer.java:90)
        at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.extractParameterInfo(EndpointIndexer.java:1251)
        at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createResourceMethod(EndpointIndexer.java:531)
        ... 14 more
```

### Expected behavior

I would have expected this to be supported. Sample code:

```java
    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello(@QueryParam("when") Optional<LocalDate> when) {
        return "Hello " + when + " from RESTEasy Reactive";
    }
```

### Actual behavior

See description.

### How to Reproduce?

Unzip [code-with-quarkus.zip](https://github.com/quarkusio/quarkus/files/9606226/code-with-quarkus.zip) and run `mvn quarkus:dev`.


### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.12.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java
+++ b/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java
@@ -146,7 +146,7 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
             DotName.createSimple("org.jboss.resteasy.reactive.server.SimpleResourceInfo"), //TODO: fixme
             RESOURCE_INFO)));
 
-    private static final Set<DotName> SUPPORT_TEMPORAL_PARAMS = Set.of(INSTANT, LOCAL_DATE, LOCAL_TIME, LOCAL_DATE_TIME,
+    protected static final Set<DotName> SUPPORT_TEMPORAL_PARAMS = Set.of(INSTANT, LOCAL_DATE, LOCAL_TIME, LOCAL_DATE_TIME,
             OFFSET_TIME,
             OFFSET_DATE_TIME, ZONED_DATE_TIME);
 
@@ -1281,8 +1281,8 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
                                 currentClassInfo, actualEndpointInfo, index);
                     }
 
-                    handleOptionalParam(existingConverters, errorLocation, hasRuntimeConverters, builder, elementType,
-                            genericElementType);
+                    handleOptionalParam(existingConverters, anns, errorLocation, hasRuntimeConverters, builder, elementType,
+                            genericElementType, currentMethodInfo);
                 }
                 builder.setOptional(true);
             } else if (convertible) {
@@ -1380,8 +1380,11 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
             boolean hasRuntimeConverters, PARAM builder, String elementType) {
     }
 
-    protected void handleOptionalParam(Map<String, String> existingConverters, String errorLocation,
-            boolean hasRuntimeConverters, PARAM builder, String elementType, String genericElementType) {
+    protected void handleOptionalParam(Map<String, String> existingConverters,
+            Map<DotName, AnnotationInstance> parameterAnnotations,
+            String errorLocation,
+            boolean hasRuntimeConverters, PARAM builder, String elementType, String genericElementType,
+            MethodInfo currentMethodInfo) {
     }
 
     protected void handleSetParam(Map<String, String> existingConverters, String errorLocation, boolean hasRuntimeConverters,

--- a/independent-projects/resteasy-reactive/server/processor/src/main/java/org/jboss/resteasy/reactive/server/processor/ServerEndpointIndexer.java
+++ b/independent-projects/resteasy-reactive/server/processor/src/main/java/org/jboss/resteasy/reactive/server/processor/ServerEndpointIndexer.java
@@ -333,8 +333,11 @@ public class ServerEndpointIndexer
         builder.setConverter(new SortedSetConverter.SortedSetSupplier(converter));
     }
 
-    protected void handleOptionalParam(Map<String, String> existingConverters, String errorLocation,
-            boolean hasRuntimeConverters, ServerIndexedParameter builder, String elementType, String genericElementType) {
+    protected void handleOptionalParam(Map<String, String> existingConverters,
+            Map<DotName, AnnotationInstance> parameterAnnotations,
+            String errorLocation,
+            boolean hasRuntimeConverters, ServerIndexedParameter builder, String elementType, String genericElementType,
+            MethodInfo currentMethodInfo) {
         ParameterConverterSupplier converter = null;
 
         if (genericElementType != null) {
@@ -350,6 +353,9 @@ public class ServerEndpointIndexer
                 converter = new SortedSetConverter.SortedSetSupplier(genericTypeConverter);
                 builder.setSingle(false);
             }
+        } else if (SUPPORT_TEMPORAL_PARAMS.contains(DotName.createSimple(elementType))) {
+            converter = determineTemporalConverter(DotName.createSimple(elementType), parameterAnnotations,
+                    currentMethodInfo);
         }
 
         if (converter == null) {
@@ -393,6 +399,11 @@ public class ServerEndpointIndexer
     protected void handleTemporalParam(ServerIndexedParameter builder, DotName paramType,
             Map<DotName, AnnotationInstance> parameterAnnotations,
             MethodInfo currentMethodInfo) {
+        builder.setConverter(determineTemporalConverter(paramType, parameterAnnotations, currentMethodInfo));
+    }
+
+    private ParameterConverterSupplier determineTemporalConverter(DotName paramType,
+            Map<DotName, AnnotationInstance> parameterAnnotations, MethodInfo currentMethodInfo) {
         String format = null;
         String dateTimeFormatterProviderClassName = null;
 
@@ -416,8 +427,7 @@ public class ServerEndpointIndexer
                         "'java.time.Instant' types must not be annotated with '@DateFormat'",
                         currentMethodInfo));
             }
-            builder.setConverter(new InstantParamConverter.Supplier());
-            return;
+            return new InstantParamConverter.Supplier();
         }
 
         if ((format != null) && (dateTimeFormatterProviderClassName != null)) {
@@ -430,23 +440,17 @@ public class ServerEndpointIndexer
         }
 
         if (LOCAL_DATE.equals(paramType)) {
-            builder.setConverter(new LocalDateParamConverter.Supplier(format, dateTimeFormatterProviderClassName));
-            return;
+            return new LocalDateParamConverter.Supplier(format, dateTimeFormatterProviderClassName);
         } else if (LOCAL_DATE_TIME.equals(paramType)) {
-            builder.setConverter(new LocalDateTimeParamConverter.Supplier(format, dateTimeFormatterProviderClassName));
-            return;
+            return new LocalDateTimeParamConverter.Supplier(format, dateTimeFormatterProviderClassName);
         } else if (LOCAL_TIME.equals(paramType)) {
-            builder.setConverter(new LocalTimeParamConverter.Supplier(format, dateTimeFormatterProviderClassName));
-            return;
+            return new LocalTimeParamConverter.Supplier(format, dateTimeFormatterProviderClassName);
         } else if (OFFSET_DATE_TIME.equals(paramType)) {
-            builder.setConverter(new OffsetDateTimeParamConverter.Supplier(format, dateTimeFormatterProviderClassName));
-            return;
+            return new OffsetDateTimeParamConverter.Supplier(format, dateTimeFormatterProviderClassName);
         } else if (OFFSET_TIME.equals(paramType)) {
-            builder.setConverter(new OffsetTimeParamConverter.Supplier(format, dateTimeFormatterProviderClassName));
-            return;
+            return new OffsetTimeParamConverter.Supplier(format, dateTimeFormatterProviderClassName);
         } else if (ZONED_DATE_TIME.equals(paramType)) {
-            builder.setConverter(new ZonedDateTimeParamConverter.Supplier(format, dateTimeFormatterProviderClassName));
-            return;
+            return new ZonedDateTimeParamConverter.Supplier(format, dateTimeFormatterProviderClassName);
         }
 
         throw new RuntimeException(

--- a/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/simple/LocalDateTimeParamTest.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/simple/LocalDateTimeParamTest.java
@@ -3,6 +3,7 @@ package org.jboss.resteasy.reactive.server.vertx.test.simple;
 import io.restassured.RestAssured;
 import java.time.LocalDateTime;
 import java.time.format.DateTimeFormatter;
+import java.util.Optional;
 import javax.ws.rs.FormParam;
 import javax.ws.rs.GET;
 import javax.ws.rs.POST;
@@ -31,6 +32,15 @@ public class LocalDateTimeParamTest {
     }
 
     @Test
+    public void localDateTimeAsOptionalQueryParam() {
+        RestAssured.get("/hello/optional?date=1984-08-08T01:02:03")
+                .then().statusCode(200).body(Matchers.equalTo("hello#1984"));
+
+        RestAssured.get("/hello/optional")
+                .then().statusCode(200).body(Matchers.equalTo("hello#2022"));
+    }
+
+    @Test
     public void localDateTimeAsPathParam() {
         RestAssured.get("/hello/1995-09-21 01:02:03")
                 .then().statusCode(200).body(Matchers.equalTo("hello@9"));
@@ -48,6 +58,12 @@ public class LocalDateTimeParamTest {
         @GET
         public String helloQuery(@RestQuery LocalDateTime date) {
             return "hello#" + date.getYear();
+        }
+
+        @Path("optional")
+        @GET
+        public String helloOptionalQuery(@RestQuery Optional<LocalDateTime> date) {
+            return "hello#" + date.orElse(LocalDateTime.of(2022, 1, 1, 0, 0)).getYear();
         }
 
         @GET

